Orc: Closes #5777 - Obtain ORC stripe offsets from writer

pavibhai opened the pull request:
## What?
Fixes ORCFileAppender such that it can retrieve the stripe offsets from the writer instead of having to reopen the file.

## Why?
From ORC 1.7 we have a public API for getting stripe information from the writer. This avoids an additional IO to get stripe offsets

## Tests
Added unit test to verify that the offsets are matching the offsets read from the file

pavibhai commented:
@kbendick Hope you are still looking at ORC support on Iceberg. Can you please take a look at this PR?

rdblue commented on orc/src/main/java/org/apache/iceberg/orc/OrcFileAppender.java:
```diff
@@ -147,11 +146,12 @@ public long length() {
     } catch (IOException e) {
-      throw new RuntimeIOException(e, "Can't close ORC reader %s", file.location());
+      throw new RuntimeIOException(
+          e, "Cannot receive stripe information from writer for %s", file.location());
```
"Cannot receive" seems like a strange way to state that `getStripes` failed. What about `Failed to get stripe information from writer for: %s`?

pavibhai replied:
Makes sense, will change.

rdblue commented on orc/src/test/java/org/apache/iceberg/orc/TestOrcDataWriter.java:
```diff
@@ -0,0 +1,276 @@
+
+  @SuppressWarnings("checkstyle:AvoidEscapedUnicodeCharacters")
+  @Test
+  public void testInvalidUpperBoundString() throws Exception {
```
Is this related to this PR? What is the purpose of this test?

pavibhai replied:
I was noticed that there was no unit test on ORC write in the project. I ported the current parquet test and introduced the check on offsets.

If you would prefer that I raise a separate PR on other unit test cases then I can split it.

pavibhai replied:
Removing the additional unit tests, will raise a separate PR for them

rdblue commented on orc/src/main/java/org/apache/iceberg/orc/OrcFileAppender.java:
```diff
@@ -147,11 +146,12 @@ public long length() {
-    try (Reader reader = ORC.newFileReader(file.toInputFile(), conf)) {
-      List<StripeInformation> stripes = reader.getStripes();
+    try {
+      List<StripeInformation> stripes = writer.getStripes();
```
This change looks great to me. Thanks for updating this, @pavibhai!

rdblue reviewed (changes requested):
Looks good other than an error message and some tests for unrelated cases.

nastra commented on orc/src/test/java/org/apache/iceberg/orc/TestOrcDataWriter.java:
```diff
@@ -0,0 +1,276 @@
+    Assert.assertEquals("Written records should match", overflowRecords, writtenRecords);
+
+    // Expect bounds on LONG column
+    Assert.assertTrue("Should have a valid lower bound", dataFile.lowerBounds().containsKey(1));
```
```suggestion
    Assertions.assertThat(dataFile.lowerBounds()).containsKey(1));
```

nastra replied:
this will show you the content of the map if the assertion ever fails, rather than just showing true != false.
would be great to apply this to the other places as well

pavibhai replied:
Thanks @nastra as suggested earlier I am removing the additional tests and will raise a separate PR. Will incorporate this feedback into that PR.

pavibhai commented:
> Looks good other than an error message and some tests for unrelated cases.

@rdblue I have addressed the review comments can you please take another look at this?

rdblue commented:
Thanks, @pavibhai! Looks great so I merged it.
